### ISSUE TITLE
configure.ac and Makefile.in hard codes tools

### ISSUE THREAD

**Issue by noloader** [NLnetLabs/nsd#85]
I'm testing NSD master on Solaris 11.3. There are several problems with compile that trace back to lack of precious variables for tools. What happens is, I set a tool like `EGREP=ggrep -E`, but the hardcoded `egrep` is used. There are two problems. First, tools like `grep` and `egrep` are not Posix; and second, Solaris managed to provide tools that are more anemic then Posix.

The fix is threefold. First, [declare the tools as precious in `configure.ac` using `AC_ARG_VAR`](https://www.gnu.org/software/autoconf/manual/autoconf-2.69/html_node/Setting-Output-Variables.html):

```
AC_ARG_VAR(SED, [location of the sed program])
AC_ARG_VAR(AWK, [location of the awk program])
AC_ARG_VAR(GREP, [location of the grep program])
...
```

Second, perform the tool tests early:

```
AC_PROG_CC
AC_PROG_SED
AC_PROG_AWK
AC_PROG_GREP
AC_PROG_EGREP
AC_PROG_LN_S
AC_PROG_INSTALL
AC_PROG_LEX
AC_PROG_YACC
```

Third, use the variables in `configure` and `Makefile`. So `Makefile.in` would be changed to:

```
SED 	= @SED@
AWK 	= @AWK@
GREP	= @GREP@
EGREP	= @EGREP@
YACC 	= @YACC@
LEX 	= @LEX@
...

head -`$(EGREP) -n "# Dependencies" $(DEPEND_TARGET) | tail -1 | $(SED) -e 's/:.*$$//'` $(DEPEND_TMP2) > $(DEPEND_TARGET)
```

After the change I was able to get through a compile on Solaris using standard tools sans one error due to an ancient version of OpenSSL.

-----

Making the tools precious variables results in the following when running `./configure --help`:

```
Some influential environment variables:
  SED         location of the sed program
  AWK         location of the awk program
  GREP        location of the grep program
  EGREP       location of the egrep program
  LEX         location of the lex program with GNU extensions (flex)
  YACC        location of the yacc program with GNU extensions (bison)
  CC          C compiler command
  ...
```

-----

Here is a difference in output between Linux and Solaris with precious variables.

**Linux**

```
$ ./configure
...
checking for a sed that does not truncate output... /bin/sed
checking for gawk... gawk
checking for grep that handles long lines and -e... /bin/grep
checking for egrep... /bin/grep -E
checking for flex... flex
...

$ make
...
/bin/sed -e 's,@prefix\@,/usr/local,g' -e 's,@exec_prefix\@,/usr/local,g' ...
```

**Solaris**

```
$ ./configure
...
checking for a sed that does not truncate output... /usr/bin/gsed
checking for gawk... gawk
checking for grep that handles long lines and -e... /usr/bin/ggrep
checking for egrep... /usr/bin/ggrep -E
checking for flex... flex
...

$ make
...
/usr/bin/gsed  -e 's,@prefix\@,/usr/local,g'  -e 's,@exec_prefix\@,/usr/local,g' ...
```

**Comment by wcawijngaards** [comment]
I appreciate your work for adding extra tests, because that is very nice.  But please don't make large edits, I mean larger than they need to be.  I do not really want to make changes that are not needed, just to increase the change volume.  We try to keep NSD minimal and lean on code to make it have less bugs and perform better.

In this case the PR looks good so I added it.  Hope that fixes the Solaris issues.  Solaris is also harder to support, since not really available, or any more.  So, if it works for your Solaris compile setup, I guess that is fine.

Thanks again, though for the added tests, we need that to find bugs.

**Comment by noloader** [comment]
@wcawijngaards ,

Oh, sorry about that. My apologies.

Solaris is getting into shape. I've got an empty `zparser.h` that is causing compile errors. I think Bison is misbehaving. I tried to build a new Bison using my [build scripts](https://github.com/noloader/Build-Scripts), but it crashes the compiler. Ugh...

I can provide SSH access to my Solaris machine. Email me your `authorized_keys` file if you are interested. Solaris is one of those OSes you want to run on real hardware. It does not play well in a VM. I run mine on a [cheap desktop](https://www.amazon.com/gp/product/B07DXRHHXK).

If you want to access Sparc hardware then I recommend [OpenCSW](https://www.opencsw.org). OpenCSW is like the MacPorts of the Solaris world. OpenCSW runs a compile farm, and provides free access to Sparc and i86pc machines running Solaris 9, 10 and 11. Also see [Upstream Maintainers](https://www.opencsw.org/extend-it/signup/to-upstream-maintainers/) on their wiki.
